### PR TITLE
[IOTDB-2871] Data node client to connect with config node

### PR DESCRIPTION
**Pull request by MrQuansy**
- ConfigNodeClient
    - Process the redirection from config node
    - Cache the leader of config node group in client

- Register data nodes using config node client
- Set global configuration after registering

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 1]
```diff
@@ -855,6 +855,17 @@ public class IoTDBConfig {
    */
   private String consensusProtocolClass = "org.apache.iotdb.consensus.ratis.RatisConsensus";
 
+  /**
+   * The series partition executor class. The Datanode should communicate with ConfigNode on startup
+   * and set this variable so that the correct class name can be obtained later when calculating the
+   * series partition
+   */
+  private String seriesPartitionExecutorClass =
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:863]: get this config from confignode.

File: server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    }
+  }
+
+  private void reconnect() throws IoTDBConnectionException {
+    if (configLeader != null) {
+      try {
+        connect(configLeader);
+        return;
+      } catch (IoTDBConnectionException e) {
+        logger.error("The current node may have been down {},try next node", configLeader);
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java:110]: logger.error might confuse user. How about logger.warn?

File: server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+      Endpoint tryEndpoint = configNodes.get(j);
+      if (j == configNodes.size() - 1) {
+        j = -1;
+      }
+      tryHostNum++;
+      try {
+        connect(tryEndpoint);
+        return;
+      } catch (IoTDBConnectionException e) {
+        logger.error("The current node may have been down {},try next node", tryEndpoint);
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java:134]: logger.error might confuse user. How about logger.warn?

File: server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+  }
+
+  public void close() {
+    transport.close();
+  }
+
+  private boolean verifyNeedRedirect(TSStatus status) throws StatementExecutionException {
+    if (status.getCode() == TSStatusCode.NEED_REDIRECTION.getStatusCode()) {
+      configLeader =
+          new Endpoint(status.getRedirectNode().getIp(), status.getRedirectNode().getPort());
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java:147]: The leader of ConfigNodeGroup is null when election. PTAL~

File: server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    }
+    throw new IoTDBConnectionException(MSG_RECONNECTION_FAIL);
+  }
+
+  public void close() {
+    transport.close();
+  }
+
+  private boolean verifyNeedRedirect(TSStatus status) throws StatementExecutionException {
+    if (status.getCode() == TSStatusCode.NEED_REDIRECTION.getStatusCode()) {
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java:145]: I set TSStatus.message when its TSStatusCode is NEED_REDIRECTION,  which you can use in logger

**Review by CRZbulabula — COMMENTED** [review]
PTAL~

File: server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java (hunk 3 above)
Comment (MrQuansy) [reply server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java:134]: Fix it.

File: server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java (hunk 4 above)
Comment (MrQuansy) [reply server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java:147]: Fix it

File: server/src/main/java/org/apache/iotdb/db/service/DataNode.java [hunk 6]
```diff
@@ -123,35 +115,34 @@ public boolean initLocalEngines() {
...
+            configNodeClient.registerDataNode(
                 new TDataNodeRegisterReq(new EndPoint(thisNode.getIp(), thisNode.getPort())));
         if (dataNodeRegisterResp.getStatus().getCode()
-            == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+                == TSStatusCode.SUCCESS_STATUS.getStatusCode()
+            || dataNodeRegisterResp.getStatus().getCode()
+                == TSStatusCode.DATANODE_ALREADY_REGISTERED.getStatusCode()) {
           dataNodeID = dataNodeRegisterResp.getDataNodeID();
+          IoTDBDescriptor.getInstance().loadGlobalConfig(dataNodeRegisterResp.globalConfig);
           logger.info("Joined a cluster successfully");
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/service/DataNode.java:138]:
```suggestion
          logger.info("Joined the cluster successfully");
```

File: server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+      }
+    }
+    throw new IoTDBConnectionException(MSG_RECONNECTION_FAIL);
+  }
+
+  public void close() {
+    transport.close();
+  }
+
+  private boolean verifyNeedRedirect(TSStatus status) throws StatementExecutionException {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/client/ConfigNodeClient.java:144]:
```suggestion
  private boolean updateConfigNodeLeader(TSStatus status) throws StatementExecutionException {
```

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**